Change #6173: Update SDL driver to use SDL 2.0

nikolas opened the pull request:
This has been working well for me. Based on the documentation here: https://wiki.libsdl.org/MigrationGuide

Will require testing on different setups.

Here's the FPS window on both old and new SDL, rendering a whole map.

SDL 1.2:
![2019-01-22-213340_1366x768_scrot](https://user-images.githubusercontent.com/59292/51579592-8a3e3500-1e8f-11e9-95d7-aec19ba781f1.png)

SDL 2.0:
![2019-01-22-214647_1366x768_scrot](https://user-images.githubusercontent.com/59292/51579610-9d510500-1e8f-11e9-80b3-8c7ec791e58c.png)

closes #7296

nikolas committed:
Change #6173: Update SDL driver to use SDL 2.0

nielsmh commented:
The number to pay attention to on those screenshots: Video output has decreased from 4 ms to 2 ms with the SDL 2 driver.
Not tested it myself.

nikolas commented:
Also worth mentioning is that this changes the SDL driver from software rendering to hardware rendering, which has to potential to cause problems since we're now relying on the user's graphics drivers.

I've seen this branch segfault when the program starts on an old Mac mini I have, running Linux and the nouveau DRI drivers. This computer can run OpenTTD with SDL 1.2 just fine, and obviously we don't want to introduce this regression.

So I'll need to implement a software rendering fallback using either SDL_CreateSoftwareRenderer() or SDL_CreateRenderer() with the SDL_RENDERER_SOFTWARE flag. I'm not sure exactly how to "fall back" to the software renderer, but I'm sure there's some examples around. I'll take a look.

nikolas commented:
Alright, I've gotten a chance to debug the issue on the Mac mini system I mentioned above. It's working fine with an update I just made to this branch: using a [renderer](https://wiki.libsdl.org/SDL_CreateRenderer) created with the `SDL_RENDERER_SOFTWARE` flag instead of `SDL_RENDERER_ACCELERATED`. Unfortunately, just passing in 0 here and letting SDL choose doesn't fix the problem: seems to be a driver bug, which other people have run into with these nouveau drivers.

Anyways, we're not losing anything by using software rendering here, as that's how it currently works. Please test this branch, anyone who's able and willing, and let me know of any problems you see!

glx22 commented:
Of course CI fails for linux because SDL2-dev is not installed

nikolas commented:
Hmm, I can't see the exact libsdl failure - though I'm not used to Azure. 

I see a bunch of messages like "Unable to find image 'openttd/compile-farm-ci:linux-amd64-gcc-6' locally" here: https://github.com/OpenTTD/OpenTTD/pull/7086/checks?check_run_id=55198664

Anyways, this Dockerfile would need to be updated to include `libsdl2-dev`. I'll make a PR. 
https://github.com/OpenTTD/CompileFarm/blob/master/base-linux/Dockerfile#L29

glx22 commented:
It fails on configure, https://github.com/OpenTTD/CompileFarm/blob/master/base-linux/Dockerfile needs an udpate

TrueBrain commented:
If you go to the Azure results, here: https://dev.azure.com/openttd/OpenTTD/_build/results?buildId=607
You can click a build that failed. Then you can click again on the step that failed. There you see: `2019-01-26T23:49:22.1701030Z configure: error: no video driver development files found`. We have an open bug to make the reporting more verbose.

PeterN commented on src/video/sdl_v.cpp:
```diff
@@ -155,19 +164,14 @@ static void DrawSurfaceToScreen()
+	SDL_CALL SDL_UpdateTexture(_sdl_texture, NULL, _sdl_screen->pixels, _sdl_screen->pitch);
+	SDL_CALL SDL_RenderClear(_sdl_renderer);
+	SDL_CALL SDL_RenderCopy(_sdl_renderer, _sdl_texture, NULL, NULL);
+	SDL_CALL SDL_RenderPresent(_sdl_renderer);
```
Updating just dirty rects was a _major_ performance improvement that definitely used to work.

I wonder why it stopped working? Our code didn't change, but OSes have... (It stopped working for the Windows driver too!)

nikolas replied:
Interesting, the [MigrationGuide](https://wiki.libsdl.org/MigrationGuide) mentions this here:

> At the end of the frame, we want to upload to the texture like this:
> 
>   SDL_UpdateTexture(sdlTexture, NULL, myPixels, 640 * sizeof (Uint32));
> 
> This will upload your pixels to GPU memory. That NULL can be a subregion if you want to mess around with dirty rectangles, but chances are modern hardware can just swallow the whole framebuffer without much trouble. The final argument is the pitch--the number of bytes from the start of one row to the next--and since we have a linear RGBA buffer in this example, it's just 640 times 4 (r,g,b,a).

When I've adapted `DrawSurfaceToScreen()` to use the dirty rectangles system, everything works, but you can actually see the dirty rectangles getting re-drawn and there is some screen flickering. There may be a way to get this working as maybe I was doing something wrong, I just haven't spent too much time on this because I wasn't sure it was necessary.

PeterN replied:
Yeah, maybe it's fine, 10+ years later? Also the GUI timer changes alleviate the load a bit.

PeterN commented on src/video/sdl_v.cpp:
```diff
@@ -155,19 +164,14 @@ static void DrawSurfaceToScreen()
+	SDL_CALL SDL_UpdateTexture(_sdl_texture, NULL, _sdl_screen->pixels, _sdl_screen->pitch);
+	SDL_CALL SDL_RenderClear(_sdl_renderer);
+	SDL_CALL SDL_RenderCopy(_sdl_renderer, _sdl_texture, NULL, NULL);
+	SDL_CALL SDL_RenderPresent(_sdl_renderer);
```
Re this change, this means that _dirty_rects will not be used, therefore it does not need to be kept around at all?

nikolas replied:
Yeah: the way this is currently working, `_dirty_rects` and `_num_dirty_rects` should just be replaced with a single `_dirty` or `_is_dirty` value (like [FreeRCT](https://github.com/FreeRCT/FreeRCT/blob/master/src/video.cpp)), to determine whether `DrawSurfaceToScreen()` should draw the entire screen or not. I'll make that change now.

PeterN commented:
Runs quite a bit slower than with SDL 1.2. I'm seeing video output averaging 2.3ms/t vs 0.8ms/t.

nikolas commented:
I am seeing a higher video output time as well when comparing scenes side-by-side. Something like 0.9ms vs 0.2ms with a certain scenario I have. I'll take another pass at getting the dirty rectangles system working on this branch.

I'm seeing a big difference in this processing time when you move the cursor rapidly. The SDL 1.2 code seems to handle this better.

LordAro commented:
@nikolas You might be interested in applying https://github.com/TrueBrain/OpenTTD/commit/647d47d3bdab0629891fe8fe76471db5d58a47cd to this PR
I suspect this one is going to land sooner than that one, anyway :)

TrueBrain commented:
Just to voice my concern here too: by removing SDL1 we might exclude some targets that are currently "supported", like MorphOS and BeOS. This might not be a bad thing, but that might mean we just remove support for those targets. There might be other unknown targets that currently only work on SDL1, and after this PR are no longer supported. Again, not a bad thing perse, but I just wanted to make sure we are all aware of this.

nikolas commented:
I've updated things here to blit the surface in exactly the same way as the SDL 1.2 driver. Video output speed is now on par with SDL 1.2 from my initial testing. Details below.

SDL 2 encourages the use of its new SDL_Texture and SDL_Renderer concepts over the older, but still available SDL_Surface concept. SDL_Texture is basically an SDL_Surface that's harder to work with, but with the advantage that all the pixels are on the GPU. This has the potential to make rendering much faster. That said, I was still forcing the renderer to use software rendering, because of the potential for segfaults on old hardware that I mentioned [here](https://github.com/OpenTTD/OpenTTD/pull/7086#issuecomment-457876339). SDL 2 doesn't provide a nice API for updating the dirty rectangles of a texture - see: https://stackoverflow.com/questions/54955273/how-do-i-use-sdl-locktexture-to-update-dirty-rectangles I finally realized I can just remove the new texture and renderer, and just deal with the window's SDL_Surface (`_sdl_realscreen`).

I've changed things to basically use SDL 2.0's legacy API: [SDL_UpdateWindowSurfaceRects()](https://wiki.libsdl.org/SDL_UpdateWindowSurfaceRects), equivalent to the old SDL_UpdateRects(), in charge of updating only the dirty rectangles. So, video output is working exactly as on SDL1 now, but if and when we ever want hardware rendering in SDL, we would need to migrate this to use SDL_Texture and SDL_Renderer.

nikolas commented:
Regarding BeOS and MorphOS - it would be a shame to remove support for operating systems that people are using. Looks like there is some incomplete SDL 2 support for MorphOS: http://www.morphos-storage.net/?find=SDL

For BeOS, it looks like if you want to use BeOS these days you would just be using HaikuOS, which does have SDL2 support. So maybe BeOS can easily be removed.. it's a proprietary OS and the latest release was in 2001.

TrueBrain commented:
I think it is fine to drop MorphOS and BeOS, especially as they didn't had any code change for the last 10 years :) No maintainers .. so no way to test anything .. so it is properly fine ;)

This is exactly what I did in #7326.

PeterN commented:
The performance regression seems to be solved now.

Resizing the window is still buggy (existing bug in SDL driver, not new to this PR). With SDL 1.2 window resizing was sort of done by recreating the game window, which some window managers didn't work very well with. I don't know if SDL 2.0 improved support for this.

PeterN commented:
Can you please also update the file `os/debian/control` on the `Build-Depends` line replace `libsdl-dev` with `libsdl2-dev`?

There may be a similar change to the rpm stuff, but I don't know anything about that. @planetmaker?

nikolas commented:
It looks like the appropriate rpm package here is [SDL2-devel](https://rpmfind.net/linux/rpm2html/search.php?query=sdl2-devel&submit=Search+...&system=&arch=), so I've updated the rpm spec as well.

PeterN commented on src/video/sdl_v.cpp:
```diff
@@ -346,39 +330,42 @@ bool VideoDriver_SDL::CreateMainSurface(uint w, uint h)
 		return false;
 	}
+
+	SDL_SetWindowSize(_sdl_window, w, h);
```
Is this call necessary? If I remove it, I can resize the game window properly, and as far as I can see nothing breaks.

EDIT: It is of course necessary when using the Game Options to change the window size. We should definitely find a way to handle this, as having working window resize is definitely a good thing.

EDIT EDIT: I've submitted a PR for this.

nikolas replied:
This seems to be necessary for changing the window size with the "screen resolution" dropdown in the game options. When I remove this, I can still resize the window by dragging the window edges via the window manager, but then the "screen resolution" dropdown doesn't change the window size.

PeterN replied:
Yup, I realised. Check out the PR -- https://github.com/nikolas/OpenTTD/pull/1

nikolas commented:
The latest commit here enables the shadow surface system for use with 8bit blitters. We'll also want to add Peter's changes here (https://github.com/PeterN/OpenTTD/commits/pr/7086) to this branch.. except for the change that removes shadow surface support.

PeterN commented:
Do you want a PR or will you cherry-pick?

nikolas commented:
I just cherry-picked your changes and added them here.

LordAro commented:
Seems to run very slowly for me on MinGW, only 30fps (goes up to 12x when fast forwarding). More curiously, it does occasionally jump up to 0.98x

![image](https://user-images.githubusercontent.com/4007992/54084025-938e3e80-4323-11e9-9a22-a3604ed36d68.png)

And I don't think this is your issue, but on MinGW this results in the following extra warnings, looks like wherever SDL.h is included:

```
<built-in>: warning: redundant redeclaration of 'unsigned int __builtin_ia32_crc32qi(unsigned int, unsigned char)' in same scope [-Wredundant-decls]
In file included from C:/msys64/mingw64/include/SDL2/SDL_cpuinfo.h:51,
                 from C:/msys64/mingw64/include/SDL2/SDL.h:38,
                 from C:/msys64/home/LordAro/OpenTTD/src/crashlog.cpp:65:
C:/msys64/mingw64/x86_64-w64-mingw32/include/intrin.h:59:23: note: previous declaration of 'unsigned int __builtin_ia32_crc32qi(unsigned int, unsigned char)'
   extern unsigned int __builtin_ia32_crc32qi (unsigned int, unsigned char);
                       ^~~~~~~~~~~~~~~~~~~~~~
<built-in>: warning: redundant redeclaration of 'unsigned int __builtin_ia32_crc32hi(unsigned int, short unsigned int)' in same scope [-Wredundant-decls]
C:/msys64/mingw64/x86_64-w64-mingw32/include/intrin.h:60:23: note: previous declaration of 'unsigned int __builtin_ia32_crc32hi(unsigned int, short unsigned int)'
   extern unsigned int __builtin_ia32_crc32hi (unsigned int, unsigned short);
                       ^~~~~~~~~~~~~~~~~~~~~~
<built-in>: warning: redundant redeclaration of 'unsigned int __builtin_ia32_crc32si(unsigned int, unsigned int)' in same scope [-Wredundant-decls]
C:/msys64/mingw64/x86_64-w64-mingw32/include/intrin.h:61:23: note: previous declaration of 'unsigned int __builtin_ia32_crc32si(unsigned int, unsigned int)'
   extern unsigned int __builtin_ia32_crc32si (unsigned int, unsigned int);
                       ^~~~~~~~~~~~~~~~~~~~~~

```

nikolas commented:
Interesting, and what video output speed and framerate are you seeing on master on this MinGW system?

nikolas commented on src/video/sdl_v.cpp:
```diff
@@ -774,10 +774,20 @@ bool VideoDriver_SDL::ToggleFullscreen(bool fullscreen)
+
+		/* Find fullscreen window size */
+		SDL_DisplayMode dm;
+		if (SDL_GetCurrentDisplayMode(0, &dm) < 0) {
```
@PeterN this section makes ToggleFullscreen work with `SDL_WINDOW_FULLSCREEN`. Previously, I was a little confused on the expected behavior here: when in windowed mode, ToggleFullscreen should actually increase the window size to be the same as the current resolution, rather than just forcing a fullscreen on the current window size.

stormcone commented:
I have the same problem as @Nik-mmzd in #7296.

At the first start after compilation the NewGRF scanning window appears and the progress bar goes up to 95% and before it closes the screen freezes. The game itself keeps running because I can hear the sounds from the title game. Just nothing happens on the screen.

After I close the game, at the following game starts the screen freezes immediately. The NewGRF scanning window appears, but the progress bar stops at 0%. The game still running in these cases as well.

My system is a Fedora 29 on a physical hardware.

Software versions:
OpenTTD: 9124b35bf81cac3c2e8222570db3840476e08e78 from your sdl2 branch
Kernel: 4.20.15-200.fc29.x86_64
SDL2: 2.0.9 (both of 32 and 64 bit versions)
NVIDIA Driver Version: 418.43

openttd -d driver=1:
dbg: [driver] Successfully loaded blitter '32bpp-anim'
dbg: [driver] SDL: using mode 1920x1054x32
dbg: [driver] SDL: using driver 'x11'
dbg: [driver] Successfully probed video driver 'sdl'
dbg: [driver] Successfully probed sound driver 'sdl'
dbg: [driver] Successfully probed music driver 'extmidi'
dbg: [driver] Successfully loaded blitter '32bpp-sse2-anim'
dbg: [driver] SDL: using mode 1920x1054x32
dbg: [driver] SDL: using threads
dbg: [driver] SDL: using mode 1920x1054x32
dbg: [driver] extmidi: set volume not implemented

LordAro commented:
Just tried to run this on Linux:

```
 > ./bin/openttd
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  151 (GLX)
  Minor opcode of failed request:  3 (X_GLXCreateContext)
  Value in failed request:  0x0
  Serial number of failed request:  109
  Current serial number in output stream:  110
```

SDL1 (master) works fine

LordAro commented:
Seems the above is due to my OpenGL being somewhat broken (it tends to happen when my Kernel/graphics driver is updated and I haven't updated yet)

But, would be nice if it could gracefully fallback like SDL1 (and other programs) do...

LordAro commented:
Yup, SDL "bug" - https://bugzilla.libsdl.org/show_bug.cgi?id=4527
Feel free to ignore the above.

LordAro commented on src/video/sdl_v.cpp:
```diff
@@ -795,29 +769,45 @@ bool VideoDriver_SDL::ChangeResolution(int w, int h)
-	bool ret = !_resolutions.empty() && CreateMainSurface(_cur_resolution.width, _cur_resolution.height);
+	/* Remember current window size */
+	if (fullscreen) {
+		SDL_GetWindowSize(_sdl_window, &_window_size_w, &_window_size_h);
```
`_window_size_*` are only used inside this function, no need for them to be global?

nikolas replied:
They don't necessarily need to be global, but they do need to be static: as it's currently set up, these variables need to exist outside the scope of this function for remembering the value of the window size throughout multiple calls of this function. Of course, this could all possibly be refactored, as maybe it's not clear enough.

glx22 commented:
A rebase is required to fix the windows linking errors (846fc8fe097689ef2fa1498708b3244fb4ad84ac)

LordAro commented:
So I actually tried this out. And... it infinite loops inside MainLoop.

As best as I can tell from some minor gdb debugging, the condition at
https://github.com/OpenTTD/OpenTTD/blob/4f0894bed3e6f6206ba7bdc4f6caba207b52927e/src/video/sdl_v.cpp#L702
never becomes true, so it spends all its time sleeping instead
What's weird is that it only stops after doing the initial NewGRF load - has it not entered MainLoop until it's finished all that?

Some sample output after putting `printf("p:%d c:%d n:%d\n", prev_cur_ticks, cur_ticks, next_tick);` just above the if condition:

![](https://i.imgur.com/EwpcLQZ.png)

nikolas commented:
Hmmmm, from that sample output, things appear to be ticking along correctly, as this line should fulfill the `cur_ticks >= next_tick` condition:
```
p:7867 c:7868 n:7868
```
Maybe some more gdb debugging could uncover more clues. That sample output is very similar to my instance which doesn't get into an infinite loop:
```
p:3122 c:3123 n:3126
p:3123 c:3124 n:3126
p:3124 c:3125 n:3126
p:3125 c:3126 n:3126
p:3126 c:3128 n:3156
p:3128 c:3132 n:3156
p:3132 c:3133 n:3156
p:3133 c:3134 n:3156
p:3134 c:3135 n:3156
p:3135 c:3137 n:3156
```

I'm trying to debug this as well, but can't reproduce the infinite loop yet. This appears to be working on some setups and not others, so finding out the difference could help us track down the real problem.

nielsmh commented:
Any chance of converting this to be a new driver alongside the existing SDL 1.2 one? It's okay to have them exclusive (cannot compile in both at the same time), but I'd at least prefer to give package maintainers the option.

nikolas commented:
Sure, I can make that change this weekend.

nikolas commented:
@nielsmh I've separated this out into a new video and sound driver. By default, the SDL 1 driver is selected as usual.

Configuring with both SDL1 and SDL2 enabled throws an error.

To compile with SDL2, you have to do: `./configure --without-sdl --with-sdl2`

I've set this up by just following what's going on in `config.lib` for the SDL configuration.

nielsmh commented:
Commit checker complains about trailing whitespace in config.lib:
`*** b/config.lib:3579: Trailing whitespace: '	echo "  --with-sdl2[=\"pkg-config sdl2\"] enables SDL2 video driver support"	'`

LordAro commented:
I'd suggest that SDL2 should be default - just fallback to SDL1 if not detected
No need for extra flags

nikolas commented:
@LordAro I've updated config.lib with your suggestions - now using SDL2 by default if it's available. You can force SDL 1 with `./configure --with-sdl=sdl1`.

glx22 commented on config.lib:
```diff
@@ -2859,7 +2874,7 @@ detect_iconv() {
 		iconv=""
 		search_paths=`LC_ALL=C $cxx_host $OSX_SYSROOT $CFLAGS -E - -v </dev/null 2>&1 | \
-		              $awk '/#include <...> search starts here:/{flag=1;next}/End of search list./{flag=0}flag'`
+					  $awk '/#include <...> search starts here:/{flag=1;next}/End of search list./{flag=0}flag'`
```
This change is unrelated to this PR

LordAro commented:
I'm struggling to work out what exactly is causing the deadlock

Some investigation shows it's not actually a deadlock - the GameLoop is running as expected, with the date continuing to tick up and the game responding to keyboard input.

DrawSurfaceToScreen is also getting called as often as expected (perhaps too often?), it's not deadlocked on the various mutexes & locks

This leaves `_sdl_surface` or `_sdl_realscreen` being at fault - the evidence seems to suggest that they're just not getting updated. Though I've no idea why it works fine up until the NewGRF load is finished, and why this is apparently working for everyone else but not me continues to be a mystery.

stormcone commented:
@LordAro You are not alone with this problem. I think I have the same issue as you as I mentioned in my comment above: https://github.com/OpenTTD/OpenTTD/pull/7086#issuecomment-474078418.
Since then I have newer system, but the problem persist with the recent updates too.

OS: Fedora 30
Kernel: 5.1.17
NVIDIA Driver Version: 430.26
SDL2: 2.0.9

nikolas commented:
LordAro is using the NVIDIA driver too - that could be the key. I don't have a computer with nvidia graphics at the moment so I can't confirm right now..

nikolas commented:
It's possible that [ShouldUseTextureFramebuffer()](https://hg.libsdl.org/SDL/file/05dddfb66b85/src/video/SDL_video.c#l541) is returning true for these nvidia situations, causing UpdateWindowFrameBuffer() to call SDL_UpdateWindowTexture(), as seen in the code linked to above. ShouldUseTextureFramebuffer() offers a few paths out of this, seen here: https://hg.libsdl.org/SDL/file/05dddfb66b85/src/video/SDL_video.c#l166

I've added a new commit to this branch that should tell SDL that we don't want any hardware acceleration, and in turn, no texture updating. We only want to be dealing with SDL surfaces, not textures, in this SDL 2 driver at the moment.

Anyways.. this is just kind of blind debugging. We'll see if this makes any difference for @LordAro or @stormcone.

stormcone commented:
@nikolas Probably you are right. I tried out your branch with a Fedora live DVD, which using the Nouveau driver, and I was able to start a new game with SDL2.

By the way is it possible to that the sdl2_v.cpp use "SDL2" in the debug messages instead of "SDL"?
For example instead of:
`DEBUG(driver, 1, "SDL: using driver '%s'", dname);`
-->
`DEBUG(driver, 1, "SDL2: using driver '%s'", dname);`

nikolas commented:
@stormcone sure, I just made that change.

nikolas commented:
@stormcone Good to hear it works with the Nouveau driver, but let me know when you're able to test this with the Nvidia proprietary driver again.

stormcone commented:
@nikolas Thank you. :)

I tried out your new commit with the disabled hardware acceleration and now the game starts normally. I can use the menu and start a new game. So looks like that commit solves the problem. :)

Edit: I mean with the NVIDIA driver.

LordAro commented:
Confirmed, now runs

LordAro commented:
Simulation rate is still consistently running at 1.02x (for me) but that's the same as SDL1, so I'm not too concerned about it

Nik-mmzd commented:
I'm on nikolas@830650615464af174f586573b01d198ba5715594; linux; sdl 2.0.9
Found two bugs:
1. Numpad Enter does not acts like Enter
2. Arrow buttons do not move cursor but ctrl-right/left arrow moves cursor by a word

Also, thanks for fixed #7296

Nik-mmzd commented:
Wow, it works, thanks

nikolas commented:
@nielsmh this is ready for another review, since known issues are now fixed. Just fyi.

LordAro commented on src/os/unix/unix.cpp:
```diff
@@ -49,7 +49,7 @@
 
 #if defined(__APPLE__)
-#	if defined(WITH_SDL)
+#	if defined(WITH_SDL) || defined(WITH_SDL2)
```
i imagine this is an SDL1-only thing, as it did funky things with main(), but SDL2 no longer does this

LordAro commented on src/video/sdl2_v.cpp:
```diff
@@ -0,0 +1,831 @@
+
+	if (bpp == 8) {
+		newscreen = SDL_CreateRGBSurfaceWithFormat(
+			0, w, h, 8, SDL_PIXELFORMAT_INDEX8);
```
weird line break here

LordAro reviewed (changes requested):
Nearly there, I think :)

LordAro reviewed (dismissed):
:+1:

TrueBrain commented on src/crashlog.cpp:
```diff
@@ -267,9 +267,15 @@ char *CrashLog::LogLibraries(char *buffer, const char *last) const
+	const SDL_version *sdl_v = SDL_Linked_Version();
+	buffer += seprintf(buffer, last, " SDL:        %d.%d.%d\n", sdl_v->major, sdl_v->minor, sdl_v->patch);
+#else
+#ifdef WITH_SDL2
```
#elif ? Otherwise this becomes a bit of a weird statement :D (and the comment on the `#endif` is in the wrong place for sure)

TrueBrain commented on src/crashlog.cpp:
```diff
@@ -267,9 +267,15 @@ char *CrashLog::LogLibraries(char *buffer, const char *last) const
-	buffer += seprintf(buffer, last, " SDL:        %d.%d.%d\n", v->major, v->minor, v->patch);
-#endif /* WITH_SDL */
+	const SDL_version *sdl_v = SDL_Linked_Version();
+	buffer += seprintf(buffer, last, " SDL:        %d.%d.%d\n", sdl_v->major, sdl_v->minor, sdl_v->patch);
```
s/SDL/SDL1/ ? For more clear error reporting?

TrueBrain reviewed (commented):
Minor, but it triggered my OCD :P
